pydis_site/apps/content/utils.py

```diff
@@ -27,8 +55,258 @@ def get_categories(path: Path) -> Dict[str, Dict]:
+                    included.append(file)
+            repo.extractall(folder, included)
+
+        for tag_file in Path(folder).rglob("*.md"):
+            name = tag_file.name
+            group = None
+            if tag_file.parent.name != "tags":
+                # Tags in sub-folders are considered part of a group
+                group = tag_file.parent.name
+
+            tags.append(Tag(
+                name=name.removesuffix(".md"),
+                sha=hashes[name],
+                group=group,
+                body=tag_file.read_text(encoding="utf-8"),
+                last_commit=None,
+            ))
+
+    return tags
+
+
+def set_tag_commit(tag: Tag) -> None:
+    """Fetch commit information from the API, and save it for the tag."""
+    if settings.STATIC_BUILD:  # pragma: no cover
+        # Static builds request every page during build, which can ratelimit it.
+        # Instead, we return some fake data.
+        tag.last_commit = Commit(
+            sha="68da80efc00d9932a209d5cccd8d344cec0f09ea",
+            message="Initial Commit\n\nTHIS IS FAKE DEMO DATA",
+            date=datetime.datetime(2018, 2, 3, 12, 20, 26, tzinfo=datetime.timezone.utc),
+            authors=json.dumps([{"name": "Joseph", "email": "joseph@josephbanks.me"}]),
+        )
+        return
+
+    path = "/bot/resources/tags"
+    if tag.group:
+        path += f"/{tag.group}"
+    path += f"/{tag.name}.md"
+
+    # Fetch and set the commit
+    with github_client() as client:
+        data = client.get("/repos/python-discord/bot/commits", params={"path": path})
+        data.raise_for_status()
+        data = data.json()[0]
+
+    commit = data["commit"]
+    author, committer = commit["author"], commit["committer"]
+
+    date = datetime.datetime.strptime(committer["date"], settings.GITHUB_TIMESTAMP_FORMAT)
+    date = date.replace(tzinfo=datetime.timezone.utc)
+
+    if author["email"] == committer["email"]:
+        authors = [author]
+    else:
+        authors = [author, committer]
+
+    commit_obj, _ = Commit.objects.get_or_create(
+        sha=data["sha"],
+        message=commit["message"],
+        date=date,
+        authors=json.dumps(authors),
+    )
+    tag.last_commit = commit_obj
+    tag.save()
+
+
+def record_tags(tags: list[Tag]) -> None:
+    """Sync the database with an updated set of tags."""
+    # Remove entries which no longer exist
+    Tag.objects.exclude(name__in=[tag.name for tag in tags]).delete()
+
+    # Insert/update the tags
+    for tag in tags:
+        try:
+            old_tag = Tag.objects.get(name=tag.name)
+        except Tag.DoesNotExist:
+            # The tag is not in the database yet,
+            # pretend it's previous state is the current state
+            old_tag = tag
+
+        if old_tag.sha == tag.sha and old_tag.last_commit is not None:
+            # We still have an up-to-date commit entry
+            tag.last_commit = old_tag.last_commit
+
+        tag.save()
```

Comment:
This part does not make sense to me. If the hash of the file changes, shouldn't we update it here? We don't seem to throw it away in that case either, but just stop updating.

Reply:
We don't fetch the hashes here as documented in the approach methodology in the PR body since it would be too expensive in terms of ratelimits. There are ~100 tags, and it doesn't make sense to make an additional 100 requests to pre-load all the commits. When someone requests a tag, that commit data is fetched and saved, otherwise no commit data will be available until then.